### PR TITLE
scheduler: hotspot: refactor data structure of store load

### PR DESCRIPTION
**Pull request by Luffbee**
### What problem does this PR solve? 

Since PR #2053 , there are two data structures to represent store load information, `ScoreInfos` and `statistics.StoreHotPeersStat`, which make things complicated.

### What is changed and how it works?

This PR is trying to use a new data structure, `map[uint64]storeLoadDetail`, to replace the original two. The new data structure has at least two advantages:
+ `storeLoadDetail` is only used internal, so we don't need to worry about compatibility in future development (comparing to `statistics.StoreHotPeersStat`).
+ `storeLoadDetail` is designed to work with "multiple dimensional load" and "load prediction" (comparing to `ScoreInfos`).

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/schedulers/hot_test.go [hunk 1]
```diff
@@ -528,8 +528,6 @@ func (s *testHotReadRegionSchedulerSuite) TestWithPendingInfluence(c *C) {
 		testutil.CheckTransferLeader(c, op1, operator.OpLeader, 1, 3)
 		op2 := hb.Schedule(tc)[0]
 		testutil.CheckTransferPeerWithLeaderTransfer(c, op2, operator.OpHotRegion, 1, 4)
-		op3 := hb.Schedule(tc)[0]
```
Comment (rleungx) [inline server/schedulers/hot_test.go]: Why remove it?

Comment (Luffbee) [reply server/schedulers/hot_test.go]: The scheduler strategy changed a little. This is a desired change.

File: server/schedulers/hot_region.go [hunk 2]
```diff
@@ -672,118 +696,118 @@ func (bs *balanceSolver) buildOperators() []*operator.Operator {
...
+			(ld1.Count == ld2.Count && ld1.ByteRate < ld2.ByteRate) {
+			return true
+		}
+		return false
+	})
+
+	srcLoad := srcLoadDetail.LoadPred.min()
+	for _, id := range stores {
+		dstLoad := candidates[id].LoadPred.max()
+		if srcLoad.Count-1 >= dstLoad.Count+1 &&
```
Comment (nolouch) [inline server/schedulers/hot_region.go]: before is `> 2` ,  and changed to `>=2`?

Comment (Luffbee) [reply server/schedulers/hot_region.go]: Yes.

File: server/schedulers/utils.go [hunk 3]
```diff
@@ -358,3 +199,76 @@ func summaryPendingInfluence(pendings map[*pendingInfluence]struct{}, f func(*op
...
+
+func (lp *storeLoadPred) min() storeLoad {
+	return min(&lp.Current, &lp.Future)
+}
+
+func (lp *storeLoadPred) max() storeLoad {
+	return max(&lp.Current, &lp.Future)
+}
+
+func min(a, b *storeLoad) storeLoad {
```
[the remark this answers is not in the document]
Comment (Luffbee) [reply server/schedulers/utils.go]: Renamed.

File: server/schedulers/utils.go [hunk 4]
```diff
@@ -358,3 +199,76 @@ func summaryPendingInfluence(pendings map[*pendingInfluence]struct{}, f func(*op
...
+
+func (lp *storeLoadPred) min() storeLoad {
+	return minLoad(&lp.Current, &lp.Future)
+}
+
+func (lp *storeLoadPred) max() storeLoad {
+	return maxLoad(&lp.Current, &lp.Future)
+}
+
+func minLoad(a, b *storeLoad) storeLoad {
```
Comment (shafreeck) [inline server/schedulers/utils.go]: Would it return a `storeLoad` with `a.ByteRate` and `b.Count`, does this work as expected?

File: server/schedulers/hot_region.go [hunk 5]
```diff
@@ -113,11 +111,9 @@ type hotScheduler struct {
 	regionPendings map[uint64][2]*operator.Operator
 
 	// temporary states but exported to API or metrics
-	stats           *storeStatistics // store id -> hot regions statistics
+	stLoadInfos     *storeLoadInfos
```
Comment (shafreeck) [inline server/schedulers/hot_region.go]: Maybe `storeLoadInfos` is more friendly?

File: server/schedulers/utils.go (hunk 4 above)
Comment (Luffbee) [reply server/schedulers/utils.go]: Yes, expected.

File: server/schedulers/hot_region.go (hunk 5 above)
Comment (Luffbee) [reply server/schedulers/hot_region.go]: That will be the same as its type. I think `stLoadInfos` is fine.

**Comment by nolouch** [comment]
PTAL @rleungx

File: server/schedulers/hot_region.go [hunk 6]
```diff
@@ -196,56 +192,47 @@ func (h *hotScheduler) dispatch(typ rwType, cluster opt.Cluster) []*operator.Ope
...
-	for storeID := range storeStatsMap {
-		if balanceType == read {
-			storeStatsMap[storeID] += h.readPendingSum[storeID].ByteRate
-		} else {
-			storeStatsMap[storeID] += h.writePendingSum[storeID].ByteRate
-		}
-	}
+func (h *hotScheduler) summaryPendingInfluence() {
+	h.readPendingSum = summaryPendingInfluence(h.readPendings, calcPendingWeight)
+	h.writePendingSum = summaryPendingInfluence(h.writePendings, calcPendingWeight)
```
Comment (rleungx) [inline server/schedulers/hot_region.go]: Is it necessary to make `calcPendingWeight` as a parameter?

File: server/schedulers/hot_region.go [hunk 7]
```diff
@@ -270,24 +257,68 @@ func (h *hotScheduler) gcRegionPendings() {
...
-func (h *hotScheduler) analyzeStoreLoad(cluster opt.Cluster, balanceType rwType) *ScoreInfos {
-	storesStats := cluster.GetStoresStats()
-	var storeStatsMap map[uint64]float64
-	if balanceType == read {
-		storeStatsMap = storesStats.GetStoresBytesReadStat()
-	} else {
-		storeStatsMap = storesStats.GetStoresBytesWriteStat()
+// Load information of all available stores.
+func summaryStoresLoad(
+	scheName string,
```
Comment (rleungx) [inline server/schedulers/hot_region.go]: remove it?

File: server/schedulers/hot_region.go (hunk 6 above)
Comment (Luffbee) [reply server/schedulers/hot_region.go]: This is not new code, it is moved here from line 881. I designed it to make `summaryPendingInfluence` more flexible. We may use some other function to calculate the weight.

File: server/schedulers/hot_region.go (hunk 7 above)
Comment (Luffbee) [reply server/schedulers/hot_region.go]: Removed.

**Pull request merged**